### PR TITLE
Add Prefix Tuning, Parallel Adapters & Mix-and-Match Adapters

### PR DESCRIPTION
**Pull request by calpt**
Depends on #289.

---

This PR adds these new adaptation methods:
- **Prefix Tuning**, proposed in [Prefix-Tuning: Optimizing Continuous Prompts for Generation (Li & Liang, 2021)](https://arxiv.org/pdf/2101.00190.pdf)
- **Parallel Adapters**, proposed in [Towards a Unified View of Parameter-Efficient Transfer Learning (He et al., 2021)](https://arxiv.org/pdf/2110.04366.pdf)
- **Mix-and-Match (MAM) Adapters**, proposed in [Towards a Unified View of Parameter-Efficient Transfer Learning (He et al., 2021)](https://arxiv.org/pdf/2110.04366.pdf)

### CODE REVIEW

**Comment by adapter-hub-bert** [comment]
This PR/issue depends on:
* Adapter-Hub/adapter-transformers#289
---
Automatically provided by **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖).

**Comment by adapter-hub-bert** [comment]
This PR/issue depends on:
* ~~Adapter-Hub/adapter-transformers#289~~
---
Automatically provided by **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖).

File: src/transformers/adapters/prefix_tuning.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+            prefix_id = self.pool.indicate_prefix(adapter_name, self.location_key)
+            self.prefixes[adapter_name] = prefix_id
+
+    def delete_adapter(self, adapter_name: str):
+        self.pool.delete_prefix(adapter_name)
+        if adapter_name in self.prefixes:
+            del self.prefixes[adapter_name]
+
+    def add_fusion_layer(self, adapter_names: Union[List, str]):
+        pass  # not applicable to prefix tuning
```
Comment (hSterz) [inline src/transformers/adapters/prefix_tuning.py:254]: Would it make sense to throw an error because this method should not be called?

**Review by hSterz — APPROVED** [review]
Looks great. Just two small questions.

**Pull request merged**